lm_eval/models/mlx_llms.py

```diff
@@ -0,0 +1,286 @@
+                full_sequence_length = non_padding_lengths[idx].item()
+                target_length = full_sequence_length - prompt_length
+                context = context_batch[idx]
+                continuation = continuation_batch[idx]
+
+                # Extract target sequence and log prob scores that correspond to the predicted probability for it
+                target_log_prob_scores = log_prob[-target_length:]
+                target_sequence = full_sequences[idx][prompt_length: full_sequence_length + 1]
+
+                # Use the target sequence for extracting log prob values from logits vocabulary distribution
+                reshaped_target_seq = mx.reshape(mx.array(target_sequence), (-1, 1))
+                target_log_probs = mx.take_along_axis(
+                    target_log_prob_scores, reshaped_target_seq, axis=1
+                )
+                # Sum over conditional log likelihood values for target sequences
+                answer_score = target_log_probs.squeeze(1).sum().item()
+
+                idx = original_order[(context, continuation)]
+
+                # Answer: (original index, log prob, is-exact-match)
+                answer = idx, answer_score, is_greedy.item()
+                res.append(answer)
+            pbar.update(1)
+
+        pbar.close()
+        # Return the answers in the original order (lost by the batch creation process)
+        return list(map(lambda i: i[1:], sorted(res, key=lambda i: i[0])))
+
+
+    def loglikelihood_rolling(
+        self, requests: list[Instance]
+    ) -> list[tuple[float, bool]]:
+        raise NotImplementedError("loglikelihood_rolling is not implemented")
+
+    def generate_until(
+        self, requests: List[Instance], disable_tqdm: bool = False
+    ) -> List[str]:
+        """
+        * Each request contains Instance.args : Tuple[str, dict] containing 1. an input string to the LM and 2. a
+          dictionary of keyword arguments used to control generation parameters.
+        * Using this input and these generation parameters, text will be sampled from the language model
+          (typically until a maximum output length or specific stopping string sequences--for example,
+          {"until": ["\n\n", "."], "max_gen_toks": 128}).
+        * The generated input+output text from the model will then be returned.
+        """
+        try:
+            from mlx_lm.utils import generate
+        except ModuleNotFoundError:
+            raise Exception(
+                "attempted to use 'mlx' LM type, but package `mlx` is not installed. Please install anthropic via "
+                "`pip install 'lm-eval[mlx]'` or `pip install -e '.[mlx]'`",
+            )
+
+        if not requests:
+            return []
+
+        res = []
+        for request in tqdm([req.args for req in requests], disable=disable_tqdm):
+            prompt, request_args = request
+            if "until" in request_args:
+                raise NotImplementedError(
```

Comment:
I think we can move this to a warning in the init (so not to flood the logger for each req), as almost all generation tasks do include a stop sequence.